src/cmd/run.go

```diff
@@ -0,0 +1,335 @@
+	}
+
+	containerConfig := containerInfo["Config"].(map[string]interface{})
+	containerState := containerInfo["State"].(map[string]interface{})
+	containerEntryPoint := containerConfig["Cmd"].([]interface{})[0]
+
+	logrus.Infof("Entry point of container '%s' is '%s'", containerName, containerEntryPoint)
+
+	if containerEntryPoint == "toolbox" {
+		logrus.Infof("Waiting for container '%s' to finish initializing", containerName)
+
+		entryPointPIDString := fmt.Sprint(containerState["Pid"])
+		entryPointPID, err := strconv.Atoi(entryPointPIDString)
+		if err != nil {
+			logrus.Debugf("PID: %s", entryPointPIDString)
+			logrus.Fatalf("Failed to parse entry point PID of container '%s'", containerName)
+		}
+
+		if entryPointPID <= 0 {
+			logrus.Debugf("PID: %d", entryPointPID)
+			logrus.Fatalf("Invalid entry point PID of container '%s'", containerName)
+		}
+
+		// Wait for initialization of a container
+		containerInitializedStamp := fmt.Sprintf("%s/container-initialized-%d", viper.GetString("TOOLBOX_RUNTIME_DIRECTORY"), entryPointPID)
+		logrus.Infof("Checking if initialization stamp %s exists", containerInitializedStamp)
+		containerInitializedTimeout := 25
+		i := 0
+		for !utils.PathExists(containerInitializedStamp) {
+			time.Sleep(time.Second)
+			i++
+
+			if i == containerInitializedTimeout {
+				logrus.Fatalf("Failed to initialize container '%s'", containerName)
+			}
+		}
+		logrus.Infof("Container '%s' is properly initialized", containerName)
+	} else {
+		logrus.Warnf("Container '%s' uses deprecated features", containerName)
+		logrus.Warn("Consider recreating it with Toolbox version 0.0.17 or newer")
+	}
+
+	logrus.Info("Checking if the current working directory is in the container")
+	args = []string{"exec",
+		"--user", viper.GetString("USER"),
+		containerName,
+		"sh", "-c", fmt.Sprintf("test -d %s", viper.GetString("PWD"))}
+	err = podman.CmdRun(args...)
+	if err != nil {
+		logrus.Fatalf("Directory '%s' does not exist in container '%s', try running 'cd && !!'", viper.GetString("PWD"), containerName)
+	}
+
+	logrus.Infof("Looking for program '%s' in container %s", commands[0], containerName)
+
+	args = []string{"exec",
+		"--user", viper.GetString("USER"),
+		containerName,
+		"sh", "-c", `command -v "$1"`, "sh", commands[0]}
+	err = podman.CmdRun(args...)
+	if err != nil {
+		if runFlags.fallbackToBash {
+			logrus.Infof("%s not found in '%s'; using /bin/bash instead", commands[0], containerName)
+			commands = nil
+			commands = []string{"/bin/bash"}
+		} else {
+			logrus.Fatalf("%s not found in '%s'", commands[0], containerName)
+		}
+	}
+
+	logrus.Infof("Running in container '%s': %v", containerName, commands)
+
+	args = []string{"exec",
+		"--user", viper.GetString("USER"),
+		"--workdir", viper.GetString("PWD")}
+
+	if !runFlags.notty {
+		args = append(args, "--interactive", "--tty")
+	}
+
+	// Add the environment variables that hold a value
+	for _, env := range preservedEnvVars {
+		value := viper.GetString(env)
```

Comment:
We should differentiate between environment variables that are unset, and those that are set to the empty string.

Reply:
Ah! That is true. I didn't realize that when I wrote it.